### PR TITLE
fix: security hardening for v1.52.0

### PR DESCRIPTION
**Pull request by erraggy**
## Summary

Comprehensive security hardening across 5 attack surfaces, addressing 26 findings from the security review at docs/plans/2026-02-20-security-hardening-design.md.

**MCP Server** (SSRF, path traversal, input validation)
- SSRF-safe HTTP client blocks requests to private/loopback IPs
- Output path sanitization prevents directory traversal via `../` in file paths
- Input validation for join spec count, collision strategies, and Go package names
- Inline content size limits and error message sanitization
- Injected SSRF-safe client for URL spec resolution

**Parser** (resource exhaustion, open redirect)
- `MaxInputSize` (default 100MB) limits `ParseReader` and `ParseBytes`
- Size-limited `fetchURL` with named constant
- Same-origin enforcement in `resolveRelativeURL` prevents open redirects via `$ref`

**HTTP Validator** (concurrency, injection, DoS)
- `sync.Map` + `atomic.Int32` pattern cache replaces `map[string]*regexp.Regexp` (race fix)
- `validationFlags` struct snapshots `StrictMode`/`IncludeWarnings` at method entry
- `url.ParseQuery` replaces manual `strings.Split` form parsing (RFC 3986 compliance)
- Fixed body presence check (removed `ContentLength == 0` false negative)
- `additionalProperties: false` enforcement in schema validation
- Error messages sanitized with `truncateForError` and `%q` formatting
- `WithMaxBodySize` option limits request/response body reads

**Generator** (code injection, path traversal)
- `toFileName` character allowlist `[a-z0-9_]` prevents path traversal via schema names
- `filepath.Base` safety in `WriteFiles` rejects names with path separators
- `cleanDescription` sanitizes spec-derived strings before embedding in Go comments
- Safer client defaults: `&http.Client{Timeout: 30 * time.Second}` replaces `http.DefaultClient`
- OAuth2/OIDC URL scheme validation with non-HTTPS warnings
- Discriminator `PropertyName` sanitization (strip quotes and backticks)

**CLI & Cross-cutting** (symlink attacks, stack overflow)
- `RejectSymlinkOutput` using `os.Lstat` before writing in fix, convert, overlay, join
- JSONPath recursive descent capped at 500 levels

Also includes go-sdk v1.3.1 update (security patch) and Go 1.24 toolchain pin for Go 1.26 compat.

## Stats
- 63 files changed, ~3976 insertions, ~164 deletions
- 8365 tests pass, `make check` clean

## Test plan
- [x] `make check` passes (lint, format, tests, coverage)
- [x] All 8365 tests pass including new security-focused tests
- [x] New tests cover: SSRF IP blocking, path traversal rejection, input validation, symlink rejection, JSONPath depth limits, pattern cache concurrency, form parsing edge cases, body presence detection, additionalProperties enforcement, error sanitization, toFileName hardening, WriteFiles safety, cleanDescription, safer client defaults

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: generator/security_gen_shared.go [hunk 1]
```diff
@@ -96,6 +97,22 @@ func addScopesFromFlow(scopeSet map[string]bool, flow *parser.OAuthFlow) {
...
+		return false
+	}
+	if u.Scheme == "https" {
+		return true
+	}
+	if u.Scheme == "http" && (u.Hostname() == "localhost" || u.Hostname() == "127.0.0.1") {
+		return true
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply generator/security_gen_shared.go]: Fixed in 3ce9821 — added `u.Hostname() == "::1"` to the localhost check.

File: parser/parser_format.go [hunk 2]
```diff
@@ -127,11 +127,15 @@ func (p *Parser) fetchURL(urlStr string) ([]byte, string, error) {
...
-	data, err := io.ReadAll(resp.Body)
+	// Read response body with size limit
+	maxSize := p.maxInputSizeOrDefault()
+	data, err := io.ReadAll(io.LimitReader(resp.Body, maxSize+1))
 	if err != nil {
 		return nil, "", fmt.Errorf("parser: failed to read response body: %w", err)
 	}
+	if int64(len(data)) > maxSize {
+		return nil, "", fmt.Errorf("parser: response body size exceeds maximum %d bytes", maxSize)
+	}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply parser/parser_format.go]: Fixed in 3ce9821 — both `parser.go` and `parser_format.go` now cap `readLimit` when `maxSize+1` overflows to negative.

File: internal/pathutil/sanitize_test.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		_, err := SanitizeOutputPath("/tmp/../etc/passwd")
+		// On most systems, filepath.Abs resolves this to /etc/passwd (no "..").
+		// The test verifies the function doesn't error for paths that Clean resolves,
+		// but to test the ".." check, we need a path where Abs doesn't fully resolve.
+		// Since most OSes resolve "..", let's just check the function works correctly:
+		// either it accepts (because Abs resolved the "..") or rejects (because ".." remains).
+		if err != nil {
+			assert.Contains(t, err.Error(), "..")
+		}
+	})
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/pathutil/sanitize_test.go]: Fixed in 3ce9821 — removed the `..` string check (see sibling comment), and updated the test to verify `filepath.Abs` resolves traversal rather than testing a dead code branch.

File: generator/oas3_generator.go [hunk 4]
```diff
@@ -980,12 +980,21 @@ func getFileDescription(name string) string {
...
-	return name
+	// Strip all characters except [a-z0-9_]
+	var result strings.Builder
+	for _, r := range name {
+		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') || r == '_' {
+			result.WriteRune(r)
+		}
+	}
+	return result.String()
 }
```
[the remark this answers is not in the document]
Comment (erraggy) [reply generator/oas3_generator.go]: Fixed in 3ce9821 — `toFileName` now returns `"unnamed"` when the allowlist strips everything.

File: parser/resolver_test.go [hunk 5]
```diff
@@ -912,6 +912,52 @@ func TestResolveHTTP_NegativeTTLDisablesCache(t *testing.T) {
...
+			if tt.wantErr {
+				require.Error(t, err)
+				assert.Contains(t, err.Error(), "cross-origin")
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, tt.wantURL, got)
+			}
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply parser/resolver_test.go]: Fixed in 3ce9821 — added 3 cross-origin rejection test cases: different host, different subdomain, and different port.

File: httpvalidator/schema.go [hunk 6]
```diff
@@ -498,17 +513,32 @@ func (v *SchemaValidator) validateFormat(s, format, path string) []ValidationErr
...
+	if v.patternCount.Add(1) > maxPatternCacheSize {
+		v.patternCache.Range(func(key, _ any) bool {
+			v.patternCache.Delete(key)
+			return true
+		})
+		v.patternCount.Store(1)
 	}
+	v.patternCache.Store(pattern, re)
 	return re.MatchString(s), nil
 }
```
[the remark this answers is not in the document]
Comment (erraggy) [reply httpvalidator/schema.go]: Won't fix — the thundering-herd window is bounded (`sync.Map.Range` is O(n) on cache size which is capped at 1000). Multiple concurrent clears waste some CPU but are correctness-safe.

File: internal/mcpserver/config.go [hunk 7]
```diff
@@ -29,6 +29,12 @@ type serverConfig struct {
 	// Validate tool defaults.
 	ValidateStrict     bool
 	ValidateNoWarnings bool
+
+	// Security settings.
+	MaxInlineSize   int64 // max inline content size in bytes (default 10 MiB)
+	MaxLimit        int   // max pagination limit (default 1000)
+	MaxJoinSpecs    int   // max specs in join (default 20)
+	AllowPrivateIPs bool  // opt-out of SSRF protection
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/mcpserver/config.go]: Won't fix — `AllowPrivateIPs` is accurate: it controls specifically whether private/loopback IPs are blocked. `DisableSSRFProtection` implies broader scope than the single mechanism it controls.

File: internal/mcpserver/server.go [hunk 8]
```diff
@@ -169,11 +173,19 @@ func makeSlice[T any](n int) []T {
 	return make([]T, 0, n)
 }
 
+// sanitizeError strips absolute filesystem paths from error messages
+// to prevent leaking internal directory structure to MCP clients.
+var pathPattern = regexp.MustCompile(`(?:/[a-zA-Z0-9._-]+){2,}`)
+
+func sanitizeError(err error) string {
+	return pathPattern.ReplaceAllString(err.Error(), "<path>")
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/mcpserver/server.go]: Won't fix — over-sanitization is the safer default for MCP error messages. Leaking internal paths is worse than occasionally mangling a `$ref` path in an error. The actual $ref resolution path is in the parsed result, not the error message.

File: internal/mcpserver/tools_join_test.go [hunk 9]
```diff
@@ -279,3 +282,123 @@ func TestJoinTool_MissingSpecInput(t *testing.T) {
...
+`, i, i, i)}
+	}
+	input := joinInput{Specs: specs}
+	result, _, err := handleJoin(context.Background(), &mcp.CallToolRequest{}, input)
+	require.NoError(t, err)
+	require.NotNil(t, result)
+	assert.True(t, result.IsError)
+	text := result.Content[0].(*mcp.TextContent).Text
+	assert.Contains(t, text, "too many specs")
+}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/mcpserver/tools_join_test.go]: Won't fix — test code panicking on wrong type is the desired behavior; it immediately surfaces the bug. Graceful failure with `require.IsType` adds verbosity without improving signal.

File: internal/mcpserver/tools_join.go [hunk 10]
```diff
@@ -48,6 +49,16 @@ func handleJoin(_ context.Context, _ *mcp.CallToolRequest, input joinInput) (*mc
...
+	if len(input.Specs) > cfg.MaxJoinSpecs {
+		return errResult(fmt.Errorf("too many specs: got %d, maximum is %d",
+			len(input.Specs), cfg.MaxJoinSpecs)), joinOutput{}, nil
+	}
+	if input.PathStrategy != "" && !validJoinStrategies[input.PathStrategy] {
+		return errResult(fmt.Errorf("invalid path_strategy: %q", input.PathStrategy)), joinOutput{}, nil
+	}
+	if input.SchemaStrategy != "" && !validJoinStrategies[input.SchemaStrategy] {
+		return errResult(fmt.Errorf("invalid schema_strategy: %q", input.SchemaStrategy)), joinOutput{}, nil
+	}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/mcpserver/tools_join.go]: Won't fix — the reviewer's analysis confirmed that `validJoinStrategies` already validates both path and schema strategies. The concern was already addressed.

File: internal/mcpserver/tools_security_test.go [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+	}
+	_, output, err := handleFix(context.Background(), &mcp.CallToolRequest{}, input)
+	require.NoError(t, err)
+	require.NotEmpty(t, output.WrittenTo)
+
+	info, err := os.Stat(outPath)
+	require.NoError(t, err)
+	// On Unix, the file mode should be 0o600 (no group/other permissions).
+	perm := info.Mode().Perm()
+	assert.Equal(t, os.FileMode(0o600), perm, "expected 0600 permissions, got %o", perm)
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/mcpserver/tools_security_test.go]: Won't fix — this project targets macOS/Linux only. Windows CI isn't configured. The test comment already notes the platform assumption.

File: internal/pathutil/sanitize_test.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+		realFile := filepath.Join(tmpDir, "real.yaml")
+		linkFile := filepath.Join(tmpDir, "link.yaml")
+
+		require.NoError(t, os.WriteFile(realFile, []byte("test"), 0o600))
+		require.NoError(t, os.Symlink(realFile, linkFile))
+
+		_, err := SanitizeOutputPath(linkFile)
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "symlink")
+	})
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/pathutil/sanitize_test.go]: Won't fix — same as #28: macOS/Linux-only project. Windows CI isn't configured.

File: internal/pathutil/sanitize.go [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+	if strings.Contains(abs, "..") {
+		return "", fmt.Errorf("pathutil: path must not contain '..': %s", abs)
+	}
+
+	info, err := os.Lstat(abs)
+	if err == nil {
+		if info.Mode()&os.ModeSymlink != 0 {
+			return "", fmt.Errorf("pathutil: refusing to write to symlink: %s", abs)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/pathutil/sanitize.go]: Acknowledged — TOCTOU is inherent to filesystem operations. The Lstat check is defense-in-depth against accidental symlink writes, not a security boundary.

File: parser/parser_options.go [hunk 14]
```diff
@@ -302,6 +304,20 @@ func WithMaxFileSize(size int64) Option {
...
+// Returns an error if size is negative.
+func WithMaxInputSize(size int) Option {
+	return func(cfg *parseConfig) error {
+		if size < 0 {
+			return fmt.Errorf("parser: maxInputSize cannot be negative")
+		}
+		cfg.maxInputSize = int64(size)
+		return nil
+	}
+}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply parser/parser_options.go]: Won't fix — `int` is the idiomatic Go parameter type for sizes that don't need to exceed platform word size. The internal representation uses `int64` but the public API uses `int` for ergonomics. This matches the existing `WithMaxFileSize` pattern established by the parser package.

File: generator/client_boilerplate.go [hunk 15]
```diff
@@ -41,7 +41,7 @@ func writeClientConstructor(buf *bytes.Buffer, info *parser.Info) {
 	buf.WriteString("func NewClient(baseURL string, opts ...ClientOption) (*Client, error) {\n")
 	buf.WriteString("\tc := &Client{\n")
 	buf.WriteString("\t\tBaseURL:    strings.TrimSuffix(baseURL, \"/\"),\n")
-	buf.WriteString("\t\tHTTPClient: http.DefaultClient,\n")
+	buf.WriteString("\t\tHTTPClient: &http.Client{Timeout: 30 * time.Second},\n")
```
[the remark this answers is not in the document]
Comment (erraggy) [reply generator/client_boilerplate.go]: Fixed in 3ce9821. Made the `time` import unconditional in both OAS3 `generateSingleClient` (`oas3_generator.go`) and OAS2 (`oas2_generator.go`) since `writeClientBoilerplate` always emits `30 * time.Second`.

File: internal/mcpserver/tools_generate_test.go [hunk 16]
```diff
@@ -174,3 +175,49 @@ func TestGenerateTool_FileInput(t *testing.T) {
...
+				text := result.Content[0].(*mcp.TextContent).Text
+				assert.Contains(t, text, "invalid package_name")
+			} else {
+				// Valid package names should not produce a package_name validation error.
+				// The result may be nil (success) or an error for a different reason.
+				if result != nil && result.IsError {
+					text := result.Content[0].(*mcp.TextContent).Text
+					assert.NotContains(t, text, "invalid package_name",
+						"valid package name %q should not be rejected", tt.pkg)
+				}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/mcpserver/tools_generate_test.go]: Won't fix. These are test assertions — if the type assertion panics, the test fails (which is the desired outcome). Adding `require.IsType` guards would just duplicate what the panic already provides. The tests verify correct behavior; incorrect types should fail loudly.

File: internal/mcpserver/tools_overlay.go [hunk 17]
```diff
@@ -91,10 +92,14 @@ func handleOverlayApply(ctx context.Context, _ *mcp.CallToolRequest, input overl
 	}
 
 	if input.Output != "" {
-		if err := os.WriteFile(input.Output, data, 0o644); err != nil {
+		cleanPath, pathErr := pathutil.SanitizeOutputPath(input.Output)
+		if pathErr != nil {
+			return errResult(fmt.Errorf("invalid output path: %w", pathErr)), overlayApplyOutput{}, nil
+		}
+		if err := os.WriteFile(cleanPath, data, 0o600); err != nil {
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/mcpserver/tools_overlay.go]: Fixed in 3ce9821. Removed the dead `strings.Contains(abs, "..")` check from `SanitizeOutputPath` — as the analysis correctly identified, `filepath.Abs` resolves all `..` components, making it dead code for traversal detection and a false-positive for legitimate paths like `/data/my..config/`. The symlink check via `os.Lstat` remains as the actual protection.

File: generator/security_gen_shared.go [hunk 18]
```diff
@@ -96,6 +97,22 @@ func addScopesFromFlow(scopeSet map[string]bool, flow *parser.OAuthFlow) {
...
+		return false
+	}
+	if u.Scheme == "https" {
+		return true
+	}
+	if u.Scheme == "http" && (u.Hostname() == "localhost" || u.Hostname() == "127.0.0.1" || u.Hostname() == "::1") {
+		return true
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply generator/security_gen_shared.go]: Won't fix. `isSecureURL` is in generated code output — it checks user-supplied OAS server URLs as strings. The three explicit hostname comparisons (`localhost`, `127.0.0.1`, `::1`) cover the practical loopback cases for URLs. Using `net.ParseIP` would add complexity for no real benefit since URLs don't contain raw IP ranges.

File: generator/security_gen_shared.go [hunk 19]
```diff
@@ -197,6 +225,12 @@ func generateOIDCDiscoveryFileShared(ctx *securityGenerationContext, schemes map
 		}
 	}
 
+	// Validate OIDC discovery URL uses HTTPS (or HTTP for localhost)
+	if discoveryURL != "" && !isSecureURL(discoveryURL) {
+		ctx.addIssue("securitySchemes.openIdConnect",
+			fmt.Sprintf("OpenID Connect discovery URL uses insecure scheme: %s", discoveryURL), SeverityWarning)
+	}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply generator/security_gen_shared.go]: Won't fix. The OIDC warning is a compile-time hint in generated code, not runtime validation. Checking the first security scheme is intentional — it covers the common single-scheme case. Multi-scheme APIs with mixed OIDC/non-OIDC configurations are rare enough to not warrant the complexity.

File: internal/mcpserver/safeclient_test.go [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+	}
+}
+
+func TestNewSafeHTTPClient(t *testing.T) {
+	client := newSafeHTTPClient()
+	require.NotNil(t, client)
+	assert.NotZero(t, client.Timeout)
+	assert.NotNil(t, client.CheckRedirect)
+	assert.NotNil(t, client.Transport)
+}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/mcpserver/safeclient_test.go]: Won't fix. The struct property test verifies the client is correctly configured. Behavioral tests (e.g. making real HTTP requests to blocked IPs) would be flaky in CI. The `isBlockedIP` function has thorough unit tests covering the actual security logic.

File: internal/mcpserver/server.go [hunk 21]
```diff
@@ -140,6 +141,9 @@ func paginate[T any](items []T, offset, limit int) []T {
 	if limit <= 0 {
 		limit = cfg.WalkLimit
 	}
+	if limit > cfg.MaxLimit {
+		limit = cfg.MaxLimit
+	}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/mcpserver/server.go]: Won't fix. `MaxLimit` has a default of 100 set in `DefaultConfig()`. A zero value means the caller explicitly configured it that way. The MCP server always uses `DefaultConfig()` as the base, so this path is unreachable in normal operation.

File: internal/pathutil/sanitize_test.go [hunk 22]
```diff
@@ -0,0 +1,93 @@
+package pathutil
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
```
[the remark this answers is not in the document]
Comment (erraggy) [reply internal/pathutil/sanitize_test.go]: Won't fix. The entire project uses testify (`assert`/`require`) consistently — it's a direct dependency in go.mod. Mixing stdlib assertions would be inconsistent.

File: parser/parser_format.go [hunk 23]
```diff
@@ -127,11 +127,20 @@ func (p *Parser) fetchURL(urlStr string) ([]byte, string, error) {
...
+	if readLimit <= 0 { // overflow: maxSize == math.MaxInt64
+		readLimit = maxSize
+	}
+	data, err := io.ReadAll(io.LimitReader(resp.Body, readLimit))
 	if err != nil {
 		return nil, "", fmt.Errorf("parser: failed to read response body: %w", err)
 	}
+	if int64(len(data)) > maxSize {
+		return nil, "", fmt.Errorf("parser: response body size exceeds maximum %d bytes", maxSize)
+	}
```
[the remark this answers is not in the document]
Comment (erraggy) [reply parser/parser_format.go]: Acknowledged. The `readLimit <= 0` guard is functionally equivalent to the suggested `maxSize < math.MaxInt64` pattern. Both prevent the CWE-190 overflow. Current form is slightly more defensive (catches any negative result, not just the MaxInt64 edge case).
